File: src/arithmetic_order_of_operations_hard.py
"""
arithmetic - order_of_operations (hard)
Generated: 2026-02-22T03:45:16.883450
"""

from problem_utils import *

def generate():
    # For harder order of operations (1600-1900), we need:
    # - Multiple nested operations
    # - Mix of operations (powers, multiplication, division, addition, subtraction)
    # - Fractions or negative numbers
    # - 4-6 operations total
    
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: Complex nested expression with powers and fractions
        # Work backwards from a clean answer
        ans = randint(1, 20)
        
        # Build: a - b * (c + d)^2 / e
        e = choice([2, 4, 5])
        inner_sum = randint(2, 5)
        c = randint(1, inner_sum - 1)
        d = inner_sum - c
        squared = inner_sum ** 2
        
        # Make b * squared / e give us something clean
        product = choice([i for i in range(2, 40) if i % e == 0])
        b = product // e
        quotient = product // e
        
        a = ans + b * squared // e
        
        expr_latex = f"{a} - {b} \\cdot ({c} + {d})^2 \\div {e}"
        
        sol = steps(
            f"First, evaluate the parentheses: ${c} + {d} = {inner_sum}$",
            f"Next, evaluate the exponent: ${inner_sum}^2 = {squared}$",
            f"Then multiply: ${b} \\cdot {squared} = {b * squared}$",
            f"Then divide: ${b * squared} \\div {e} = {quotient}$",
            f"Finally subtract: ${a} - {quotient} = {ans}$"
        )
        
        return problem(
            question=f"Evaluate: ${expr_latex}$",
            answer=ans,
            difficulty=(1600, 1750),
            topic="arithmetic/order_of_operations",
            solution=sol
        )
    
    elif problem_type == 2:
        # Type 2: Multiple nested parentheses with mixed operations
        # [(a + b) * c - d] / e + f
        ans = randint(5, 25)
        f_val = randint(1, 10)
        before_add = ans - f_val
        
        e = choice([2, 3, 4, 5])
        numerator = before_add * e
        
        d = randint(1, 10)
        before_subtract = numerator + d
        
        c = choice([2, 3, 4, 5])
        inner_sum = before_subtract // c if before_subtract % c == 0 else (before_subtract // c + 1)
        
        if before_subtract % c != 0:
            c = choice([i for i in [2, 3, 4, 5] if before_subtract % i == 0])
            inner_sum = before_subtract // c
        
        a = randint(1, inner_sum - 1)
        b = inner_sum - a
        
        expr_latex = f"\\frac{{({a} + {b}) \\cdot {c} - {d}}}{{{e}}} + {f_val}"
        
        sol = steps(
            f"Evaluate innermost parentheses: ${a} + {b} = {inner_sum}$",
            f"Multiply: ${inner_sum} \\cdot {c} = {before_subtract}$",
            f"Subtract: ${before_subtract} - {d} = {numerator}$",
            f"Divide: ${numerator} \\div {e} = {before_add}$",
            f"Add: ${before_add} + {f_val} = {ans}$"
        )
        
        return problem(
            question=f"Evaluate: ${expr_latex}$",
            answer=ans,
            difficulty=(1650, 1800),
            topic="arithmetic/order_of_operations",
            solution=sol
        )
    
    elif problem_type == 3:
        # Type 3: Expression with multiple exponents and operations
        # a^2 + b * c^2 - d^2
        ans = randint(10, 50)
        
        d = randint(2, 5)
        d_sq = d ** 2
        
        before_subtract = ans + d_sq
        
        c = randint(2, 4)
        c_sq = c ** 2
        
        b = choice([i for i in range(1, 6) if (before_subtract - i * c_sq) > 0])
        a_sq = before_subtract - b * c_sq
        
        # Find if a_sq is a perfect square
        a = int(a_sq ** 0.5)
        if a * a != a_sq:
            a = randint(2, 5)
            a_sq = a ** 2
            before_subtract = a_sq + b * c_sq
            ans = before_subtract - d_sq
        
        expr_latex = f"{a}^2 + {b} \\cdot {c}^2 - {d}^2"
        
        sol = steps(
            f"Evaluate all exponents first: ${a}^2 = {a_sq}$, ${c}^2 = {c_sq}$, ${d}^2 = {d_sq}$",
            f"Multiply: ${b} \\cdot {c_sq} = {b * c_sq}$",
            f"Add: ${a_sq} + {b * c_sq} = {before_subtract}$",
            f"Subtract: ${before_subtract} - {d_sq} = {ans}$"
        )
        
        return problem(
            question=f"Evaluate: ${expr_latex}$",
            answer=ans,
            difficulty=(1600, 1750),
            topic="arithmetic/order_of_operations",
            solution=sol
        )
    
    else:
        # Type 4: Complex fraction with operations in numerator and denominator
        # (a * b + c^2) / (d - e)
        ans = randint(5, 20)
        
        d = randint(10, 20)
        e = randint(1, d - 2)
        denominator = d - e
        
        numerator = ans * denominator
        
        c = randint(2, 5)
        c_sq = c ** 2
        
        product = numerator - c_sq
        if product <= 0:
            c = 2
            c_sq = 4
            product = numerator - c_sq
        
        b = choice([i for i in [2, 3, 4, 5] if product % i == 0])
        a = product // b
        
        expr_latex = f"\\frac{{{a} \\cdot {b} + {c}^2}}{{{d} - {e}}}"
        
        sol = steps(
            f"Evaluate the numerator: ${a} \\cdot {b} = {product}$",
            f"Evaluate the exponent: ${c}^2 = {c_sq}$",
            f"Complete numerator: ${product} + {c_sq} = {numerator}$",
            f"Evaluate the denominator: ${d} - {e} = {denominator}$",
            f"Divide: ${numerator} \\div {denominator} = {ans}$"
        )
        
        return problem(
            question=f"Evaluate: ${expr_latex}$",
            answer=ans,
            difficulty=(1700, 1850),
            topic="arithmetic/order_of_operations",
            solution=sol
        )

emit(generate())

File: src/arithmetic_order_of_operations_hard.jl
# arithmetic - order_of_operations (hard)
# Generated: 2026-03-08T19:51:07.068187

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("arithmetic/order_of_operations")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Complex nested expression with exponents and multiple operations
        a = randint(2, 5)
        b = randint(2, 4)
        c = randint(3, 7)
        d = randint(2, 6)
        e = randint(10, 25)
        f = randint(2, 8)
        
        ans = a^b * (c + d) - e * f
        expr_str = "$(a)^{$(b)} \\cdot ($(c) + $(d)) - $(e) \\cdot $(f)"
        
        problem(
            question="Evaluate: \$$(expr_str)\$",
            answer=ans,
            difficulty=(1800, 2000),
            solution=steps(
                "First evaluate the exponent: \$$(a)^{$(b)} = $(a^b)\$",
                "Evaluate parentheses: \$($(c) + $(d)) = $(c + d)\$",
                "Perform multiplications: \$$(a^b) \\cdot $(c + d) = $(a^b * (c + d))\$ and \$$(e) \\cdot $(f) = $(e * f)\$",
                "Subtract: \$$(a^b * (c + d)) - $(e * f) = $(ans)\$"
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Fraction with nested operations and exponents
        a = randint(2, 4)
        b = randint(2, 3)
        c = randint(5, 12)
        d = randint(2, 8)
        e = randint(2, 5)
        f = randint(1, 4)
        
        numerator = a^b + c * d
        denominator = e^f
        ans = numerator // denominator
        
        expr_str = "\\frac{$(a)^{$(b)} + $(c) \\cdot $(d)}{$(e)^{$(f)}}"
        
        problem(
            question="Evaluate and simplify: \$$(expr_str)\$",
            answer=ans,
            difficulty=(1850, 2050),
            solution=steps(
                "Evaluate numerator exponent: \$$(a)^{$(b)} = $(a^b)\$",
                "Multiply in numerator: \$$(c) \\cdot $(d) = $(c * d)\$",
                "Add numerator: \$$(a^b) + $(c * d) = $(numerator)\$",
                "Evaluate denominator: \$$(e)^{$(f)} = $(denominator)\$",
                sol("Answer", ans)
            ),
            time=150
        )
        
    elseif problem_type == 3
        # Multiple nested parentheses with exponents
        a = randint(2, 4)
        b = randint(3, 8)
        c = randint(2, 5)
        d = randint(4, 9)
        e = randint(2, 6)
        f = randint(2, 3)
        
        ans = (a * (b - c) + d)^f - e
        expr_str = "\\left($(a) \\cdot ($(b) - $(c)) + $(d)\\right)^{$(f)} - $(e)"
        
        problem(
            question="Evaluate: \$$(expr_str)\$",
            answer=ans,
            difficulty=(1900, 2100),
            solution=steps(
                "Innermost parentheses: \$($(b) - $(c)) = $(b - c)\$",
                "Multiply: \$$(a) \\cdot $(b - c) = $(a * (b - c))\$",
                "Add inside brackets: \$$(a * (b - c)) + $(d) = $(a * (b - c) + d)\$",
                "Apply exponent: \$($(a * (b - c) + d))^{$(f)} = $((a * (b - c) + d)^f)\$",
                "Subtract: \$$(( a * (b - c) + d)^f) - $(e) = $(ans)\$"
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Complex expression with division and multiple operations
        a = randint(3, 8)
        b = randint(2, 5)
        c = randint(6, 15)
        d = randint(2, 4)
        e = randint(3, 9)
        f = randint(2, 6)
        g = randint(10, 30)
        
        ans = a * b^d + c ÷ e - f * g
        expr_str = "$(a) \\cdot $(b)^{$(d)} + $(c) \\div $(e) - $(f) \\cdot $(g)"
        
        problem(
            question="Evaluate: \$$(expr_str)\$",
            answer=ans,
            difficulty=(2000, 2200),
            solution=steps(
                "Evaluate exponent: \$$(b)^{$(d)} = $(b^d)\$",
                "Perform multiplications and division left to right: \$$(a) \\cdot $(b^d) = $(a * b^d)\$, \$$(c) \\div $(e) = $(c ÷ e)\$, \$$(f) \\cdot $(g) = $(f * g)\$",
                "Add and subtract left to right: \$$(a * b^d) + $(c ÷ e) = $(a * b^d + c ÷ e)\$",
                "Final subtraction: \$$(a * b^d + c ÷ e) - $(f * g) = $(ans)\$"
            ),
            time=150
        )
        
    elseif problem_type == 5
        # Triple nested expression with exponents
        a = randint(2, 4)
        b = randint(2, 3)
        c = randint(5, 12)
        d = randint(2, 7)
        e = randint(3, 8)
        f = randint(2, 5)
        g = randint(1, 3)
        
        ans = a^b * ((c - d) * e + f^g)
        expr_str = "$(a)^{$(b)} \\cdot \\left(($(c) - $(d)) \\cdot $(e) + $(f)^{$(g)}\\right)"
        
        problem(
            question="Evaluate: \$$(expr_str)\$",
            answer=ans,
            difficulty=(2100, 2300),
            solution=steps(
                "Evaluate exponents: \$$(a)^{$(b)} = $(a^b)\$ and \$$(f)^{$(g)} = $(f^g)\$",
                "Innermost parentheses: \$($(c) - $(d)) = $(c - d)\$",
                "Multiply: \$($(c - d)) \\cdot $(e) = $((c - d) * e)\$",
                "Add inside brackets: \$$(( c - d) * e) + $(f^g) = $((c - d) * e + f^g)\$",
                "Final multiplication: \$$(a^b) \\cdot $((c - d) * e + f^g) = $(ans)\$"
            ),
            time=180
        )
        
    else
        # Advanced: Expression with fraction bar acting as grouping
        a = randint(2, 5)
        b = randint(3, 8)
        c = randint(2, 6)
        d = randint(4, 10)
        e = randint(2, 4)
        f = randint(3, 7)
        g = randint(2, 3)
        
        numerator = a * b^g - c * d
        denominator = e + f
        ans = numerator // denominator
        
        expr_str = "\\frac{$(a) \\cdot $(b)^{$(g)} - $(c) \\cdot $(d)}{$(e) + $(f)}"
        
        problem(
            question="Evaluate and simplify: \$$(expr_str)\$",
            answer=ans,
            difficulty=(2200, 2400),
            solution=steps(
                "Evaluate exponent in numerator: \$$(b)^{$(g)} = $(b^g)\$",
                "Perform multiplications in numerator: \$$(a) \\cdot $(b^g) = $(a * b^g)\$ and \$$(c) \\cdot $(d) = $(c * d)\$",
                "Subtract in numerator: \$$(a * b^g) - $(c * d) = $(numerator)\$",
                "Add in denominator: \$$(e) + $(f) = $(denominator)\$",
                "Divide: \$\\frac{$(numerator)}{$(denominator)} = $(ans)\$"
            ),
            time=180
        )
    end
end